features/map.go

```diff
@@ -0,0 +1,173 @@
+package features
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+type MapCache struct {
+	mu       sync.RWMutex
+	mapTypes map[ebpf.MapType]error
+}
+
+var (
+	mc MapCache
+)
+
+func init() {
+	mc.mapTypes = make(map[ebpf.MapType]error)
+}
+
+// FlushMapCache invalidates the entire cache storing feature probe results.
+func FlushMapCache() {
+	mc.mu.Lock()
+	// could this approach introduce any unwanted side effects in multiple threads
+	// even if we lock access to the cache structure?
+	// should I rather delete() all the entries in the map?
+	mc.mapTypes = make(map[ebpf.MapType]error)
+	mc.mu.Unlock()
+}
+
+// FlushMapCacheEntry allows to delete a specified entry of the MapType feature cache.
+func FlushMapCacheEntry(mt ebpf.MapType) {
+	mc.mu.Lock()
+	delete(mc.mapTypes, mt)
+	mc.mu.Unlock()
+}
+
+func probeMapTypeAttr(mt ebpf.MapType) *internal.BPFMapCreateAttr {
+	var (
+		keySize               uint32 = 4
+		valueSize             uint32 = 4
+		maxEntries            uint32 = 1
+		innerMapFd            uint32
+		flags                 uint32
+		btfKeyTypeID          uint32
+		btfValueTypeID        uint32
+		btfFd                 uint32
+		btfVmLinuxValueTypeID uint32
+	)
+
+	// switch on map types to generate correct bpfMapCreateAttr
+	switch mt {
+	case ebpf.StackTrace:
+		// valueSize needs to be sizeof(uint64)
+		valueSize = 8
+	case ebpf.LPMTrie:
+		keySize = 8
+		valueSize = 8
+		flags = unix.BPF_F_NO_PREALLOC
+	case ebpf.ArrayOfMaps:
+		fallthrough
+	case ebpf.HashOfMaps:
+		innerMapFd = ^uint32(0)
+	case ebpf.CGroupStorage:
+		fallthrough
+	case ebpf.PerCPUCGroupStorage:
+		// Why struct{u32 + u64} = 12 + padding = 16 ? can it be 12 for 32-bit cpus?
+		keySize = 16
+		valueSize = 8
+		maxEntries = 0
+	case ebpf.Queue:
+		fallthrough
+	case ebpf.Stack:
+		keySize = 0
+	case ebpf.StructOpts:
+		// we can't support StructOps probes currently as it will require a valid BTF fd
+		btfVmLinuxValueTypeID = 1
```

Comment:
Thanks @qmonnet, that's useful! Lorenz suggested this might be possible. Which BTF blob are you using there? I've spun this out into https://github.com/cilium/ebpf/issues/344 for now.

Reply:
:+1: I pasted the complete patch on the new issue.

Reply:
Thanks! One more thing, you mentioned:

> The `btfVmLinuxValueTypeID = 1` is wrong anyway

Why is this? Is this a hardcoded check in the kernel? According to https://www.kernel.org/doc/html/v5.9/bpf/btf.html#type-encoding, type ID 3 is `BTF_KIND_ARRAY`, at least as far as vmlinux is concerned. I think for eBPF programs, the compiler can (and does) assign type IDs 1 and up to any arbitrary type.

If this is really needs to be `BTF_KIND_ARRAY`, we should probably use constants for this.

Reply:
My understanding of BTF is not perfect, but the value we want here is the BTF id corresponding to `struct bpf_struct_ops_tcp_congestion_ops` in kernel's BTF. The value I needed to have it work was `26843`, as readable from:

```
# bpftool btf dump id 1 | grep bpf_struct_ops_tcp_congestion_ops
[26843] STRUCT 'bpf_struct_ops_tcp_congestion_ops' size=256 vlen=3
```

I'm not confident at all that this can be a constant. My guess is that this is fixed for a given kernel BTF object, but ids would probably be numbered differently on a kernel with different definitions.

It doesn't _have_ to point to the BTF of the valid struct (I think?), but if we want to create the map we have to use a `type_id` from which the `st_ops` retrieved with `bpf_struct_ops_find_value()` passes the different checks in `bpf_struct_ops_map_alloc()` in kernel/bpf/bpf_struct_ops.c. I spent some time trying to forge something that would work, but eventually the solution I have in the patch above was way cleaner (and the only one I managed to make work).

Reply:
> My understanding of BTF is not perfect, but the value we want here is the BTF id corresponding to struct bpf_struct_ops_tcp_congestion_ops in kernel's BTF. The value I needed to have it work was 26843

Yes, type IDs are dynamic and will differ between binaries. (The same is true for the size btw, which I see you've hardcoded to `value_size = 256` in your example in the other issue. The struct size can be pulled from the BTF too.)

> I spent some time trying to forge something that would work, but eventually the solution I have in the patch above was way cleaner (and the only one I managed to make work).

Great, thanks for the research! If successfully creating any StructOps map depends on vmlinux BTF being available, then I'm afraid there's no way around it. We can try and fiddle around a bit more, but we'll probably end up querying the BTF tree in this library as well, then.